WASM tail calls should work in OMG tier

justinmichaud opened the pull request:
#### a2db78b8fdbef3caa09cb0165315ef552a073590
<pre>
WASM tail calls should work in OMG tier
<a href="https://bugs.webkit.org/show_bug.cgi?id=273875">https://bugs.webkit.org/show_bug.cgi?id=273875</a>

Reviewed by Keith Miller.

Tom&apos;s original patch was correct, but we told him to elide moves to temporary storage when they
weren&apos;t needed. Unfortunately, the current version on ToT clobbers a bunch of important stuff. Let&apos;s
fix that.

We also fix up LLInt, since it had a bit of bit rot.

Finally, we add a new kind of tail call, a fake tail call, for the case when an inlined call makes
a tail call.

There is still more work to be done to ensure that this code is correct; Tests for stack traces,
inlined calls that clobber instance, and more advanced stack / regalloc situations are all
still needed.

* JSTests/wasm/stress/simd-tail-call-simple.js: Added.
* JSTests/wasm/stress/tail-call-js.js: Added.
* JSTests/wasm/stress/tail-call-simple.js: Added.
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:

Canonical link: <a href="https://commits.webkit.org/281110@main">https://commits.webkit.org/281110@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/27b7bd41c72bc44d15f29421affe58f001ca3c9a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58640 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37967 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11125 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62267 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9084 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45604 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9282 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47457 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6470 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60671 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35539 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50704 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28309 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32283 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8005 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8088 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/51731 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54235 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8277 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63970 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/57882 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2553 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8281 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54779 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2562 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50730 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54863 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12993 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2146 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/79643 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33796 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13275 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34882 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35966 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34627 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -298,25 +298,13 @@ macro restoreCalleeSavesUsedByWasm()
-    else
-        error
-    end
+    storep gpr0, CodeBlock[sp]
+    storep gpr1, ArgumentCountIncludingThis[sp]
```
What is the reason of this change?

justinmichaud replied:
I changed the slot that got saved to (callee -> ACIT), so it can't be an stp anymore. That is because callee needs to be set by the caller now

Constellation commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -298,25 +298,13 @@ macro restoreCalleeSavesUsedByWasm()
-        error
-    end
+    loadp CodeBlock[sp], gpr0
+    loadp ArgumentCountIncludingThis[sp], gpr1
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -1617,8 +1661,14 @@ auto OMGIRGenerator::addCrash() -> PartialResult
 {
-    bool isTailCall = callType == CallType::TailCall;
-    ASSERT(callType == CallType::Call || isTailCall);
+    const bool isFakeTailCall = callType == CallType::TailCall && m_inlineParent;
```
Let's name it TailCallInlineCaller etc., which is aligned to DFG / FTL.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -1660,34 +1710,35 @@ auto OMGIRGenerator::emitIndirectCall(Value* calleeInstance, Value* calleeCode,
+            // PC has already been pushed to the stack, so the stack pointer looks a bit weird.
+            offset += sizeof(Register);
+#endif
+            jit.storeWasmCalleeCallee(params[patchArgsIndex + 1].gpr(), offset);
```
Use `prologueStackPointerDelta` abstraction, which is telling how many bytes we will change in the prologue.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4682,34 +4742,383 @@ auto OMGIRGenerator::createCallPatchpoint(BasicBlock* block, Value* jsCalleeAnch
+            jit.store64(tmp, dst);
+            jit.load64(src.withOffset(bytesForWidth(Width::Width64)), tmp);
+            jit.store64(tmp, dst.withOffset(bytesForWidth(Width::Width64)));
+        }
```
If you can use scratchRegister here, then you can use `transfer32`, `transfer64`.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4682,34 +4742,383 @@ auto OMGIRGenerator::createCallPatchpoint(BasicBlock* block, Value* jsCalleeAnch
+    JIT_COMMENT(jit, "SP[", safeAreaLowerBound, "] to SP[", stackUpperBound, "] form the safe portion of the stack to clobber; Scratches go from SP[0] to SP[", scratchAreaUpperBound, "].");
+
+    // FIXME: this should be debug only, but I am leaving it in for now to make fuzzing easier.
+    // Clobber all safe values to make debugging easier.
```
Let's wrap it with
```
#if ASSERT_ENABLED
#endif 
```

kmiller68 replied:
Yeah, could be wrapped with ASSERT_ENABLED as Yusuke recommends.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4682,34 +4742,383 @@ auto OMGIRGenerator::createCallPatchpoint(BasicBlock* block, Value* jsCalleeAnch
+#if CPU(X86_64)
+    // On intel, we need to push the return PC too.
+    newSPAtPrologueOffsetFromSP -= sizeof(Register);
+#endif
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4682,34 +4742,383 @@ auto OMGIRGenerator::createCallPatchpoint(BasicBlock* block, Value* jsCalleeAnch
+    }
+#endif
+
+    for (unsigned i = 2; i < 50; ++i) {
+        // Everthing after sp might be overwritten anyway.
+        jit.store64(MacroAssembler::TrustedImm32(0xBFFF), CCallHelpers::Address(MacroAssembler::stackPointerRegister, -i * sizeof(uint64_t)));
+    }
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4682,34 +4742,383 @@ auto OMGIRGenerator::createCallPatchpoint(BasicBlock* block, Value* jsCalleeAnch
+#if CPU(X86_64)
+            // We already pushed the return address
+            newFP++;
+#endif
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4723,59 +5132,67 @@ auto OMGIRGenerator::createTailCallPatchpoint(BasicBlock* block, CallInformation
-        block->appendNew<B3::MemoryValue>(m_proc, B3::Store, Origin(), value, address);
-    };
+    ASSERT(wasmCalleeInfoAsCallee.params.size() == tmpArgSourceLocations.size());
+    for (unsigned i = 0; i < patchArgs.size(); ++i) {
+        // We will clobber our stack, so we shouldn't be reading any special extra patch args from it after this point.
+        ASSERT(patchArgs[i].rep().isReg() || patchArgs[i].rep().isConstant());
+        ASSERT(!scratchRegisters.contains(patchArgs[i].rep().reg(), IgnoreVectors));
+    }
```
Let's wrap it with ASSERT_ENABLED.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4852,8 +5267,10 @@ auto OMGIRGenerator::emitInlineDirectCall(uint32_t calleeFunctionIndex, const Ty
 {
-    bool isTailCall = callType == CallType::TailCall;
-    ASSERT(callType == CallType::Call || isTailCall);
+    const bool isFakeTailCall = callType == CallType::TailCall && m_inlineParent;
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4924,8 +5332,14 @@ auto OMGIRGenerator::addCall(uint32_t functionIndex, const TypeDefinition& signa
+#if CPU(X86_64)
+                    // PC has already been pushed to the stack, so the stack pointer looks a bit weird.
+                    thisSlotOffset += sizeof(Register);
+#endif
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
```diff
@@ -4970,15 +5393,14 @@ auto OMGIRGenerator::addCall(uint32_t functionIndex, const TypeDefinition& signa
+            // PC has already been pushed to the stack, so the stack pointer looks a bit weird.
+            if (isTailCall)
+                offset += sizeof(Register);
+#endif
```
Ditto.

kmiller68 reviewed (commented):
Overall, looks fine but you missed some ASSERT_ENABLEDs Yusuke asked for.

kmiller68 reviewed (approved):
r=me.

justinmichaud committed:
WASM tail calls should work in OMG tier
https://bugs.webkit.org/show_bug.cgi?id=273875

Reviewed by Keith Miller.

Tom's original patch was correct, but we told him to elide moves to temporary storage when they
weren't needed. Unfortunately, the current version on ToT clobbers a bunch of important stuff. Let's
fix that.

We also fix up LLInt, since it had a bit of bit rot.

Finally, we add a new kind of tail call, a fake tail call, for the case when an inlined call makes
a tail call.

There is still more work to be done to ensure that this code is correct; Tests for stack traces,
inlined calls that clobber instance, and more advanced stack / regalloc situations are all
still needed.

* JSTests/wasm/stress/simd-tail-call-simple.js: Added.
* JSTests/wasm/stress/tail-call-js.js: Added.
* JSTests/wasm/stress/tail-call-simple.js: Added.
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:

Canonical link: https://commits.webkit.org/281110@main
